xformers/benchmarks/benchmark_mem_eff_attention.py

```diff
@@ -5,172 +5,236 @@
-            grad = torch.ones_like(q)
-
-            torch.cuda.reset_peak_memory_stats()
-            torch.cuda.synchronize()
-            results.append(
-                benchmark.Timer(
-                    stmt="out.backward(grad, retain_graph=True)",
-                    globals={
-                        "out": out,
-                        "grad": grad,
-                    },
-                    label="attention",
-                    description="optimized",
-                    sub_label=sub_label,
-                    num_threads=num_threads,
-                ).blocked_autorange(min_run_time=min_run_time)
-            )
-            torch.cuda.synchronize()
-            memory = torch.cuda.max_memory_allocated() / 2**20
-            mem_use["optimized"][sub_label] = memory
-            memory_str = f"Memory used: {memory} MB"
-
-            print("Optimized", memory_str)
-
-            out = ref_attention(q, q, q)
-            torch.cuda.reset_peak_memory_stats()
-            torch.cuda.synchronize()
-            results.append(
-                benchmark.Timer(
-                    stmt="out.backward(grad, retain_graph=True)",
-                    globals={
-                        "out": out,
-                        "grad": grad,
-                    },
-                    label="attention",
-                    description="vanilla",
-                    sub_label=sub_label,
-                    num_threads=num_threads,
-                ).blocked_autorange(min_run_time=min_run_time)
-            )
-
-            torch.cuda.synchronize()
-            memory = torch.cuda.max_memory_allocated() / 2**20
-            mem_use["vanilla"][sub_label] = memory
-            memory_str = f"Memory used: {memory} MB"
-            print("Vanilla", memory_str)
-
-    compare = benchmark.Compare(results)
-    compare.print()
-
-    pprint.pprint(mem_use)
-
-
-benchmark_forward()
-benchmark_backward()
+
+def benchmark_forward(shape, num_threads: int, attn_bias_type, dtype):
+    B, M, K = shape
+    q = torch.rand(shape, device=device, dtype=dtype)
+
+    dispatch = xformers.ops.AttentionOpDispatch(
+        dtype=dtype,
+        device=device,
+        k=K,
+        attn_bias_type=attn_bias_type,
+        has_dropout=False,
+        kv_len=M,
+        q_len=M,
+    )
+    try:
+        op = dispatch.op if FORCE_OP is None else FORCE_OP
+    except NotImplementedError:
+        return
+    if not op.supports(dispatch):
+        return
+
+    attn_bias = create_attn_bias(
+        attn_bias_type,
+        batch_size=B,
+        q_len=M,
+        kv_len=M,
+        device=device,
+        dtype=dtype,
+    )
+
+    dtype_str = {
+        torch.bfloat16: "b16",
+        torch.half: "f16",
+        torch.float: "f32",
+    }[dtype]
+    sub_label = f"{dtype_str} {op.NAME} B={B}, M={M}, K={K}"
+
+    if True:
+        r = xformers.ops.memory_efficient_attention(q, q, q, attn_bias, op=op).float()
```

Comment:
nit: I would be curious to get numbers on self vs. non-self attention, if anything I think that benchmarking your work in the self attention case (while very relevant for vision of course) sells you short, since there's more opportunity for the GPU to have a hot cache (and the vanilla computation is IO bottlenecked, so will benefit a lot from that)